test/upgrade/upgrade.go

```diff
@@ -17,37 +17,197 @@ limitations under the License.
 package upgrade
 
 import (
+	"knative.dev/serving/pkg/apis/autoscaling"
+	"knative.dev/serving/test/e2e"
 	"testing"
 
 	// Mysteriously required to support GCP auth (required by k8s libs).
 	// Apparently just importing it is enough. @_@ side effects @_@.
 	// https://github.com/kubernetes/client-go/issues/242
 	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
 
-	pkgTest "knative.dev/pkg/test"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ptest "knative.dev/pkg/test"
+	revisionresourcenames "knative.dev/serving/pkg/reconciler/revision/resources/names"
+	serviceresourcenames "knative.dev/serving/pkg/reconciler/service/resources/names"
+	v1a1testing "knative.dev/serving/pkg/testing/v1alpha1"
+	v1b1testing "knative.dev/serving/pkg/testing/v1beta1"
 	"knative.dev/serving/test"
 	v1a1test "knative.dev/serving/test/v1alpha1"
+	v1b1test "knative.dev/serving/test/v1beta1"
 )
 
+// These service names need to be stable, since we use them across
+// multiple "go test" invocations.
 const (
-	// These service names need to be stable, since we use them across
-	// multiple "go test" invocations.
-	serviceName            = "pizzaplanet-upgrade-service"
-	scaleToZeroServiceName = "scale-to-zero-upgrade-service"
+	ServiceName            = "pizzaplanet-upgrade-service"
+	ScaleToZeroServiceName = "scale-to-zero-upgrade-service"
 )
 
 // Shamelessly cribbed from conformance/service_test.
 func assertServiceResourcesUpdated(t *testing.T, clients *test.Clients, names test.ResourceNames, routeDomain, expectedText string) {
 	t.Helper()
 	// TODO(#1178): Remove "Wait" from all checks below this point.
-	_, err := pkgTest.WaitForEndpointState(
+	_, err := ptest.WaitForEndpointState(
 		clients.KubeClient,
 		t.Logf,
 		routeDomain,
-		v1a1test.RetryingRouteInconsistency(pkgTest.MatchesAllOf(pkgTest.IsStatusOK, pkgTest.EventuallyMatchesBody(expectedText))),
+		v1a1test.RetryingRouteInconsistency(ptest.MatchesAllOf(ptest.IsStatusOK, ptest.EventuallyMatchesBody(expectedText))),
 		"WaitForEndpointToServeText",
 		test.ServingFlags.ResolvableDomain)
 	if err != nil {
 		t.Fatalf("The endpoint for Route %s at domain %s didn't serve the expected text \"%s\": %v", names.Route, routeDomain, expectedText, err)
 	}
 }
+
+//CreateService creates a new service and test if it is created successfully using a given
+//version of client
+func CreateService(t *testing.T, apiVersion string) {
+	clients := e2e.Setup(t)
+
+	var names test.ResourceNames
+	names.Service = ServiceName
+	names.Image = test.PizzaPlanet1
+	var domain string
+	switch apiVersion {
```

Comment:
This switch statement is going to be difficult to maintain. This would be good to just refactor into separate functions

Reply:
done